gateway_pp_test.go

```diff
@@ -0,0 +1,228 @@
+package infrared
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/haveachin/infrared/protocol"
+	"github.com/haveachin/infrared/protocol/handshaking"
+	"github.com/pires/go-proxyproto"
+)
+
+// Checking or proxy is working by checking or the remoteAddr's port is different from the port the client is using
+func getIpFromAddr(addr net.Addr) string {
+	return strings.Split(addr.String(), ":")[0]
+}
+
+func createConnWithFakeIP(gatewayAddr string) (Conn, error) {
+	dialer := &net.Dialer{
+		LocalAddr: &net.TCPAddr{
+			IP:   net.ParseIP("127.0.10.1"),
+			Port: 0,
+		},
+	}
+	netConn, _ := dialer.Dial("tcp", gatewayAddr)
+	return wrapConn(netConn), nil
+}
+
+func TestProxyProtocolOn(t *testing.T) {
+	serverAddr := "infrared.gateway"
+	listenAddr := ":26572"
+	gatewayPort := 25572
+	gatewayAddr := fmt.Sprintf(":%d", gatewayPort)
+
+	cDial := make(chan error)
+	cListen := make(chan error)
+	cGateway := make(chan error)
+	cResult := make(chan bool)
+	cInfo := make(chan string)
+
+	go func() {
+		config := &ProxyConfig{
+			DomainName:    serverAddr,
+			ListenTo:      gatewayAddr,
+			ProxyTo:       listenAddr,
+			ProxyProtocol: true,
+		}
+
+		var proxies []*Proxy
+		proxy := &Proxy{Config: config}
+		proxies = append(proxies, proxy)
+		gateway := Gateway{}
+
+		if err := gateway.ListenAndServe(proxies); err != nil {
+			cGateway <- err
+			return
+		}
+	}()
+
+	go func() {
+		listener, err := Listen(listenAddr)
+		if err != nil {
+			cListen <- err
+			return
+		}
+		defer listener.Close()
+
+		proxyListener := &proxyproto.Listener{Listener: listener.Listener}
+		defer proxyListener.Close()
+
+		conn, err := proxyListener.Accept()
+		if err != nil {
+			cListen <- err
+			return
+		}
+		defer conn.Close()
+
+		ip := getIpFromAddr(conn.RemoteAddr())
+
+		cInfo <- ip
+
+	}()
+
+	go func() {
+		time.Sleep(time.Second * 1) // Startup time for gateway
+		conn, err := createConnWithFakeIP(gatewayAddr)
+		if err != nil {
+			cDial <- err
+			return
+		}
+		defer conn.Close()
+
+		isHandshakePk := handshaking.ServerBoundHandshake{
```

Comment:
Long variable names that exist over a short scope shouldn't have such long names. I would suggest to just call it 'hs' for **h**and**s**hake since it's kind of obvious in this context.